### ISSUE TITLE
Implement pouchdb-find for idb-next

### ISSUE THREAD

**Issue by daleharvey** [pouchdb/pouchdb#6543]
I wanted to open a discussion thread about this as its long been the main point and blocker for releasing `idb-next`

**Comment by daleharvey** [comment]
Ok, so this is one of the main reasons I wanted to do `idb-next` as it would be possible to implement `pouchdb-find` but not based on top of mapreduce. This could potentially fix a huge amount of performance and size issues in PouchDB.

Now onto implementation details, for one this will likely mean a change in syntax for `pouchdb-find`, in `next` we have a single objectstore with a 1:1 mapping of documents, so we can create an indexeddb index on that objectstore and use that for queries, the main issue is that currently pouchdb-find does `createIndex` and indexeddb requires indexes to be defined as part of the schema

I suspect we want this to look something like 

```
var db = new PouchDB('games', { 
  indexes: [{fields: ['name']}]
});

return db.find({
  selector: {name: {$gt: null}},
  sort: ['name']
});
```

(hopefully `indexes` could look a little nicer, just guessing from the example @ http://nolanlawson.github.io/pouchdb-find/)

**Comment by garrensmith** [comment]
Using the built in indexing would be great. Is it possible to create indexes later at all?

**Comment by daleharvey** [comment]
It would be super hard, even maintaining the indexes is going to be tricky, indexeddb requires them to be defined as part of the schema so each time an index changes it will also be a schema change (that we need to make sure we can also handle the typical schema upgrades). Its possible to do that at runtime, but it sounds extremely messy and likely to breakages.

**Comment by garrensmith** [comment]
Ah ok, I remember we had a discussion around this before. Is there no way for us to do some kind of bitwise incrementing or something every time we need to add/remove an index? This isn't going to be easy but we do need a decent way to add/remove indexes

**Comment by daleharvey** [comment]
oh yup sorry, we will definitely be able to add/remove them, just only on construction, so we do 

```
var db = new PouchDB('games', { 
  indexes: [{fields: ['name']}]
});
```
and on startup we can check if there is a name index (or any unused indexes) and create / delete them, I only think its too complicated to do them at runtime (ie `db.createIndex`) but we defintely need to be able to add / remove them.

**Comment by daleharvey** [comment]
Trying to find the original discussion about this but not finding it yet

**Comment by nolanlawson** [comment]
Maybe one of these?

- https://github.com/pouchdb/pouchdb/issues/3775
- https://github.com/pouchdb/pouchdb/issues/5216
- https://github.com/pouchdb/pouchdb/issues/2280

BTW I think your new plan to includes the indexes in the constructor makes sense. We don't have a choice unless we want to build our own secondary index system on top of IDB's so that we can give users the ability to define indexes on-the-fly. (E.g. create one objectStore called `secondary-index`, then use stringification and prefixing to create indexes on-the-fly.)

In previous discussions, one reason I was in favor of that approach was to support multiEntry/arraykeys/etc. in IE and Edge, but by the time this feature is done IE11 will probably be dipping under 1% market share and dropping like a stone, so that just leaves Edge, but using pure native IDB indexes will give my team a healthy incentive to implement the feature, and plus it'll be less JS to ship to users, so I'm :+1:. :smiley:

**Comment by SCdF** [comment]
Hey @nolanlawson / @daleharvey I and others at @medic would love to pick this up, as efficiency is becoming a bigger and bigger deal for us.

I'm trying to get up to speed with the various attempts and tickets. Based on #5207 it looks like this is the current existing implementation of this idea: https://github.com/pouchdb/pouchdb/tree/master/packages/node_modules/pouchdb-adapter-indexeddb

Is that still relevant? Is it that this probably works, but pouchdb-find needs to be modified to support it?

**Comment by SCdF** [comment]
According to @daleharvey sounds like I'm on the right track: https://github.com/pouchdb/pouchdb/issues/3775#issuecomment-266892683

> [idb-next is] @ #5207, its basically the pouchdb-find implementation that needs done then to worry about migrations and verify performance etc. Been hammered at work recently but hoping to get on it soon, very happy if anyone else wants to take a look

Looks like the discussion on implementation took place initially on #4984, but was moved and concluded on #5216. Then implementation continued against #5207. It's not clear that anything has been done against pouchdb-find specifically yet, at least nothing that has made it into master.

**Comment by daleharvey** [comment]
Yup exactly, https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-adapter-indexeddb/src/index.js needs `.find` implemented, ideally we could reuse the entire current pouchdb-find test suite however the syntax (construction) i going to be different so may be tricky, I would copy and paste a very basic subset of tests that cover most of the core functionality, get that working then we can worry about the full scale test suite

**Comment by daleharvey** [comment]
Hopefully I have written down enough of my thoughts about this in various bugs, but if you have any more questions, irc is not so common for me these days but will make sure to answer in here

**Comment by nename0** [comment]
While doing #6858, I noticed that the 'next' `indexeddb` adapter does not use getAll()/getAllKeys() in the allDocs function. Is this intended or just something more to be implemented(probably mostly copy-past from `idb`) ?

**Comment by daleharvey** [comment]
Yup to be implemented, I remember getAll / getAllKeys took a while to get implemented / full support, I think this was done before that

**Comment by nerumo** [comment]
dear bot, please don't :)

**Comment by SCdF** [comment]
So I'm pretty new to indexeddb in general, but I wrote [a small benchmark page](https://scdf.github.io/idbbench/index.html) to look into the performance differences between:
 - Current PouchDB (using IndexedDB and custom views)
 - Direct IndexedDB with one native view per PouchDB view
 - Direct IndexedDB with one artificial view, with the keyrange scoped to the name of the PouchDB view

As there was some concern noted that having an artificial view would not necessarily have great performance characteristics. 

The code is not exactly award-winning, and comparing direct IDB with PouchDB is wildly unfair (in the idb impls I just write the doc, I'm sure PouchDB is supporting _changes and all sorts of other great stuff), but I think it's interesting in a couple of ways, at least to me:
 - Direct idb doesn't require view warming, which is a pretty massive performance saving AFAICT.
 - While the artificial view concept is slower than the native one, it's still way faster than native PouchDB in raw view queries.

Testing with 50 randomly generated views and 10,000 documents (this is not an unrealistic load for us) (in millis, on a MBP 13" 2015 Chrome): 
```
{
  "insert-pouchdb": 7932770,
  "insert-native": 79656,
  "insert-artificial": 76654,
  "warm-pouchdb": 524523,
  "query-pouchdb": 248,
  "query-native": 31,
  "query-artificial": 41
}
```

So PouchDB takes roughly 100x to insert, though I'm sure 10x of that is stuff that idbnext would need anyway. There is ~9 minutes of view warming which presumably just doesn't apply natively. And then querying is ~5-10x faster regardless of implementation, which is hopefully mostly retained once it's integrated into the PouchDB plugin system

**Comment by daleharvey** [comment]
Will work on this, promise

**Comment by SCdF** [comment]
@daleharvey FWIW we at @medic have this in our longer-term milestones to work on, which we might hit near the middle of this year, unless you get to it first :-)

We will be updating this ticket though, if we start actively working on it.

**Comment by daleharvey** [comment]
Awesome thats great to hear, will also make sure to keep up to date if / when any progress is made, will be plenty of work for all :)

**Comment by garrensmith** [comment]
I think we want to keep this open

________________________________

**Comment by SCdF** [comment]
@daleharvey we at @medic (me specifically) have got to our longer-term milestones noted above, and so are starting to actively look into what needs to be done here

**Comment by daleharvey** [comment]
Awesome, so yeh my main issue has been figuring out where to start, basically can / should we base this on top of abstract mapreduce, reuse a bunch of code, rework the tests to use the constructor. Or is all that code deeply based on assumptions that dont hold for native indexes and would it be harder to reuse it than to rewrite it all. 

I suspect we can reuse it, maybe having it reworked a little, but I dont know that side of the codebasae very well, feel free to start on any code and if I get anything done I will make a PR out of it and we can collaborate on that if that sounds good

**Comment by garrensmith** [comment]
From my experience with CouchDB Mango and PouchDB-Find. There is 3 sections for this we need to look at:

1. How do we define an index. From my understanding we need to use the constructor for that. And then how would we add more.
2. When a user makes a query we need to choose the correct index and the correct query key parameters. I'm happy to work on this part and I think we need to possibly look at rewriting this from scratch. But can take parts of PouchDB-find.
3. Once an index is selected we need to get the values from it and then do further processing of the results with the selector to remove any docs from the index that might not match the selector. This we should be able to use PouchDB-find for we should just clean up the api around that part so we just do something like `pouchdbFind.match(selector, doc)`

Does that help and make sense? I'm happy to go into more detail on this.

**Comment by SCdF** [comment]
Awesome, thanks for the background folks.

FWIW my experience here is very close to zero, we just really need to up PouchDB's performance on slow phones / larger datasets, and this seems like an obvious thing to crack on and improve. So I apologise in advance for the inevitably dumb questions I will have along the way (and my speed in making changes in an entirely unfamiliar codebase).

My first port of call is to look at both the old codebase for this datastore and the new one to work out how they are different, as well as work out how to get a build that [our application](/medic/medic-webapp) can deploy with (because I want to be able to benchmark / feel out "real life" performance issues, especially on old devices).

At this stage I'm unsure what has changed (with idb) that would make the new idb adapter faster than the old one: I've heard things that imply mapreduce may be identical and find / mango are where the improvements lie. Except if that were true we probably wouldn't need an entirely new adapter? I'm going to give the code a read now and see if I can intuit some things.

I've also [played around very briefly with indexeddb](https://github.com/SCdF/idbbench/blob/master/index.html#L280-L299) and, if the problem is that we have our own index implementations, I *think* we could use a "compound index" (apologies for my probably incorrect terminology) for mapreduce view emits, and potentially individual indexes for mango. Though using individual indexes comes with the issues noted by @garrensmith in point 1 above: because you need to close the DB when creating indexes it doesn't seem practical to parse a ddoc and realise you need to add new indexes.

It seems reasonably realistic though, to have one native compound index which is basically `[indexName, whateverYouEmitBeItASingleValueOrArray]`.

Anyway, I'm going to dig around and hopefully learn some things!

**Comment by garrensmith** [comment]
Wow @SCdF this is amazing work. A really nice code example to look through. I like the idea of using the last 900 to support pouchdb versions. We don't bump that version number that often. In terms of a datetime versus a long I don't mind either. 

I'm not sure about the idea of checking `_design/docs` for indexes. What about rather doing something like this for a first attempt:

```js
var db = new PouchDB('database', {indexes: ["name", ["name", "age"], "location"]})
```

That way the indexes are defined on startup. Rather than us opening and closing the db often on the discovery of design docs. This would require a user to know beforehand what indexes to use and it would deviate from CouchDB but it should be a safer way of doing it. 

Thanks for working on this. It's really exciting to see the progress you making.

**Comment by birtles** [comment]
Thank you so much to both of you for your work on this!

Whilst you are discussing API design I wanted to comment that I hope whatever approach this takes, that it will be possible to support array keys eventually (i.e. IndexedDB's [multiEntry](https://w3c.github.io/IndexedDB/#index-multientry-flag) feature).

Recently I find many indexes I create end up using array keys (tags on documents are a particularly common use case).

**Comment by SCdF** [comment]
@birtles can you currently define multi-array keys in the existing Mango index api (my familiarity is building, but I'm not 100% yet)?

Or are you suggesting that this is an extension to the spec that PouchDB could support at some point?

**Comment by birtles** [comment]
> @birtles can you currently define multi-array keys in the existing Mango index api (my familiarity is building, but I'm not 100% yet)?

Mongo DB automatically detects array fields and [generates a multikey index](https://docs.mongodb.com/manual/core/index-multikey/).

For Mango, however, [this thread](https://issues.apache.org/jira/browse/COUCHDB-2867) describes how Couch currently requires you to add `.[]` as a suffix to the array field, but also notes that it shouldn't be necessary to do so.

**Comment by garrensmith** [comment]
@SCdF Sorry I didn't mean to put a spanner in the works. I'm ok with either way of defining an index, I was just proposing another way i.e some bad bikeshedding. Ideally, we just need to test to make sure there are no major performance issues or usage issues with creating an index. So if we want to look at each design doc and then open/close a database to create an index we just need to see if that causes any noticeable performance issue. Also what happens if a user is trying to create and read documents at the same time as this is happening.

**Comment by SCdF** [comment]
No probs. I think the two issues you raise should be fine, specifically though the second one needs to be soak tested to be sure. It would make sense though, since it appears the main point of the event driven framework is to account for this kind of thing (cross tab / transaction madness being dealt with cleanly)

**Comment by SCdF** [comment]
Update: working through this, as I'm building up compatibility with queries between CouchDB and IndexedDB I've run into some issues.

IndexedDB's native indexes **do not support indexing documents when the value to be indexed is `null`, `true` and `false`** (CouchDB does).

This presents as the index ignoring the document (no new index entry is created), and attempting to create `IDBKeyRange` documents that reflect these values (e.g. `IDBKeyRange.only(false)`) throws an exception.

There may be other edge cases that I have yet to come across.

Specifically, [idb supports the following keys](https://www.w3.org/TR/IndexedDB/#key-construct):
![image](https://user-images.githubusercontent.com/583851/47491551-faaec900-d842-11e8-893c-e0838215e1db.png)

So far I think there are two obvious solutions for this, and I'm interested in people's thoughts:
 1. Don't worry about being compatible with CouchDB, and just support what indexeddb supports. We would need to clearly document the differences, including the less obvious silent inverted ones (if you query for `{foo: {$ne: 'bar'}}` you won't get instances where `foo` is `null`, `true` or `false` because that document wouldn't be represented in the index). Due to a lack of types we also couldn't intelligently fall back onto the old form of indexing either, because you never know when someone will put an unsupported value into that json path.
 2. Worry about being compatible with CouchDB, and modify the data in such a way before storage so we can index it, being careful to revert it back to its initial values whenever the user retrieves it. For example, the boolean `false` might be stored instead as the string `"_pouch_boolean:false"`. We would need to update / override the [existing clone function](https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-utils/src/clone.js) to support this, as well as making sure all places that put and get full documents run through the same logic, and then knowing about these conventions when querying (eg querying for `$eq: false` eventually becomes `IDBKeyRange.only('_pouch_boolean:false')`). Since we're already cloning and inspecting all the values there would be no real additional performance cost to this, it's just a tiny bit funky, and I'm not 100% sure it would work in all cases yet.

I am personally heavily leaning toward 2., because we're already doing similar stuff (moving `_id` and `_rev` to `id` and `rev`, the existing clone function etc),  I don't see it creating any performance issues, and CouchDB compatibility (with good performance) is a huge deal for me.

Interested in your thoughts!

**Comment by SCdF** [comment]
Bad stalebot!

**Comment by ptitjes** [comment]
@daleharvey Will we have a `7.1` version to test that and also have the many authentication bug fixes ?

**Comment by garrensmith** [comment]
Woohoo well done 🎉


________________________________
